### PR TITLE
Ghostferry idempotence

### PR DESCRIPTION
**Pull request by fjordan**
## Overview
This PR adds both integration tests and TLA+ spec modifications that explicitly verify the idempotence of ghostferry's algorithm.

### Integration Tests
Several tests were added to the ruby integration test suite that perform combinations of interrupting, resuming, and interrupting or failing again before running finally to completion. These tests utilize assertions that the source and target are equal and some assertions that inspect the logs to ensure it indeed ran the intended number of times.

### TLA+ Spec
Additions to the TLA+ specification (via Pluscal) were added and checked with TLC to verify the correctness of these changes. 

To model the idempotence of the `TableIterator`, an `either...or` clause in the `tblit_upkey` label was added that may or may not increment the `PK` of the `TableIterator` as it is iterating the data of the source table. The effect of this is that the current row may be applied more than one time as the data is being copied.

Likewise, a `label` was added to the `BinlogWriter` (`reapply_binlog_write`) that duplicates the first entry to the `BinlogWriteQueue`. The effect of this is that any binlog entry may be applied more than once, but while also preserving the order of entries in this queue.

Following the other simplifications and proofs by induction in the spec, we make the simplification that replaying the current row in the `TableIterator` and current entry in the `BinlogStreamer` is the same as replaying the last _N_ rows or operations (in order).

Also note that while making these changes, I was able to violate the `SourceTargetEquality` condition in a TLC run by copying the current first element of the `BinlogWriteQueue` to the _end_ of the queue instead of placing it again at the _beginning_. This proves further that, as we have modeled in the implementation, the ordering of these events must remain the same and that we cannot apply them arbitrarily, and as long as they are applied in order, idempotence does indeed hold. 

### Why?
See https://github.com/Shopify/ghostferry/issues/200 for details.

/cc @Shopify/pods

### CODE REVIEW

File: binlog_streamer.go [hunk 1]
```diff
@@ -369,11 +368,6 @@ func (s *BinlogStreamer) handleRowsEvent(ev *replication.BinlogEvent, query []by
 		}
 
 		events = append(events, dmlEv)
-		s.logger.WithFields(logrus.Fields{
```
Comment (fjordan) [inline binlog_streamer.go:372]: This was removed as it was spamming the logs and preventing the travis tests from completing. I don't think it adds much value

**Comment by fjordan** [comment]
After discussions with @shuhaowu, I've decided to rework the TLA+ changes proposed in this PR to model the idempotence of the algorithm.

**Comment by milanatshopify** [comment]
> After discussions with @shuhaowu, I've decided to rework the TLA+ changes proposed in this PR to model the idempotence of the algorithm.

Can we do "both"?  Or is that just a waste of computers?

**Comment by shuhaowu** [comment]
We're still going to keep the tests here right? They looked decent as far as I remember.

**Comment by fjordan** [comment]
> Can we do "both"? Or is that just a waste of computers?

We want to make sure we don't move the spec further away from the actual implementation, which is what the additions here were unfortunately doing. We could keep them, because yes, they do model idempotence in one way. I think we'd prefer the alternative with something like multiple initial starting states.

> We're still going to keep the tests here right? They looked decent as far as I remember.

Yes. I'll break them out into a separate PR.

**Pull request closed without merging**